### ISSUE TITLE
[Dlang] `d_module_versions` is not applied if passed in `declare_dependency`

### ISSUE THREAD

**Issue by denizzzka** [mesonbuild/meson#7168]
**Describe the bug**
`d_module_versions` is not applied if passed in `declare_dependency`

**Sample:**
```Meson
druntime_dep = declare_dependency(
    sources: [
        freertos_druntime_backend,
    ],
    include_directories: include_directories(
        './freertos_druntime_backend/',
    ),
    dependencies: [
        druntime_dep_orig,
        freertos_d_dep,
        picolibc_dep,
    ],
    d_module_versions: [ #this section will be ignored
        'DruntimeAbstractRt',
        'CRuntime_Abstract',
        'ExternStackShell',
        'AsmExternal',
    ],
)
```

**system parameters**
* Is this a [cross build](https://mesonbuild.com/Cross-compilation.html) or just a plain native build (for the same computer)?
cross build
* what `meson --version`
current ~master 589a6249f0ffb0295c3f15233d1b6b3af9e4de16

**Comment by denizzzka** [comment]
Probably, related to #6821

**Comment by denizzzka** [comment]
```d
// D language test code:
// dmd -version=CustomVersion1 -version=CustomVersion2 bug_versions.d && ./bug_versions # returns 0
// dmd -version=CustomVersion1 bug_versions.d && ./bug_versions # returns 1
// dmd -version=CustomVersion2 bug_versions.d && ./bug_versions # returns 1
// dmd bug_versions.d && ./bug_versions # returns 1
int main()
{
    // Test d_module_versions CustomVersion1 and CustomVersion2
    version(CustomVersion1)
    {
        version(CustomVersion2)
        {
            return 0;
        }
    }

    return 1;
}
```

**Comment by dcbaker** [comment]
Would it make sense to remove the `d_unittest` flag and handle unittests the way we handle them in rust, with a function that essentially clones the target, then adds the necessary test arguments? https://mesonbuild.com/Rust-module.html#functions

**Comment by dcbaker** [comment]
Part of my rational is to reduce the explosion of options we pass to `build_target` (and it's wrappers like library, executable, etc). Part is that it makes it simple to build both a test version and a version suitable for install).

It's been a while since I looked at D, but last time I looked an executable built with `-unittest` would always run the unittests when run, sounds like that's changed?

**Comment by denizzzka** [comment]
> It's been a while since I looked at D, but last time I looked an executable built with -unittest would always run the unittests when run, sounds like that's changed?

Yes, 2-3 years ago this behaviour was changed. Now you can run only unittests or main code:

```D
import std.stdio;

void main()
{
    writeln("main body");
}

unittest
{
    writeln("unit test");
}
```

```
$ dmd sample.d && ./sample
main body
$ dmd -unittest sample.d && ./sample
unit test
```

In old days a unittest version would have returned:
```
$ dmd -unittest sample.d && ./sample
unit test
main body
```

**Comment by rtbo** [comment]
This issue was addressed in #10061.
`d_module_versions` and `d_import_dirs` have been added to `declare_dependency`.
Is it worth adding more?

**Comment by denizzzka** [comment]
> d_module_versions and d_import_dirs have been added to declare_dependency.

Works! Great! Thank you!

> Is it worth adding more?

I think remaining two options (`d_debug list` and `d_unittest`) is not worth. Perhaps someone has a different opinion, but it can be described in a another issue
